### PR TITLE
fix: stream confusion matrix in val loop to prevent CPU OOM on large datasets

### LINKED ISSUE #717: train_segmentation_landcover: CPU OOM after epoch 1 on large datasets (val loop accumulates full tensor)
## Bug

`train_segmentation_landcover` (via `_train_with_custom_iou`) crashes with SIGKILL at end of epoch 1 on large datasets. Occurs on memory-limited HPC nodes (SLURM, V100 32 GB GPU).

## Root cause

The validation loop accumulates the **entire validation set** as `int64` CPU tensors each epoch:

```python
all_preds = []
all_targets = []
with torch.no_grad():
    for images, targets in val_loader:
        ...
        preds = torch.argmax(outputs, dim=1).cpu()
        all_preds.append(preds)
        all_targets.append(targets.cpu())
all_preds = torch.cat(all_preds, dim=0)
all_targets = torch.cat(all_targets, dim=0)
```

**Memory math** (256×256 tiles, int64 = 8 bytes):
- Per tile: `256 × 256 × 8 × 2` ≈ 1 MB
- 30k–80k val tiles → **30–80 GB CPU RAM** at epoch 1 boundary
- Plus `pin_memory=True` locked RAM + worker fork CoW → exceeds SLURM `--mem` → SIGKILL

## Reproducer

Any training run where val set exceeds ~10k tiles:
```python
train_segmentation_landcover(
    images_dir=...,
    labels_dir=...,  # e.g. 50k tiles total, val_split=0.2 → 10k val tiles
    ...
)
# Killed at end of epoch 1
```

## Fix

Stream a **confusion matrix** (O(num_classes²) memory) for `validation_iou_mode="mean"` instead of collecting full tensors. Other modes retain existing behavior.

PR: https://github.com/jayakumarpujar/geoai-1/pull/new/fix/landcover-val-oom

## Environment

- geoai 0.32.0
- PyTorch 2.x, CUDA 12.x
- SLURM HPC, V100 32 GB GPU, `--mem=64G`

### PR DESCRIPTION
**Pull request by jayakumarpujar**
Fixes #717

## Summary

- Validation loop in `_train_with_custom_iou` accumulated the entire validation set as `int64` CPU tensors each epoch (`all_preds` / `all_targets`), causing 30–80 GB peak RAM on large datasets and SIGKILL on HPC nodes
- Replace with streaming confusion matrix (`O(num_classes²)` memory) for the default `validation_iou_mode='mean'` path
- Non-mean modes (`perclass_frequency`, `boundary_weighted`, `sparse_labels`) retain existing behavior

## Changes

`geoai/landcover_train.py` — `_train_with_custom_iou` validation block:
- Detect `stream_iou = (iou_mode == "mean")` before val loop
- In stream mode: accumulate `num_classes × num_classes` confusion matrix on CPU per batch (no tensor lists)
- Compute mIoU from confusion matrix diag (`tp / (tp + fp + fn)`)
- Guard `isinstance(ignore_index, bool)` so `False` (no-ignore) is not treated as class 0
- Exclude `ignore_index` from `present` mask to match `landcover_iou(mode='mean')` behaviour

## Test plan

- [ ] Run training on 50k+ tile dataset with `--mem=64G` SLURM node — epoch 1 must complete without SIGKILL
- [ ] `seff <jobid>` MaxRSS stays < 30 GB regardless of dataset size
- [ ] Streaming mIoU matches `landcover_iou(mode='mean')` numerically on small dataset (including ignored pixels)
- [ ] All existing modes (`perclass_frequency`, `boundary_weighted`, `sparse_labels`) pass existing tests

### CODE REVIEW

**Comment by jayakumarpujar** [comment]
Closing in favour of clean PR from #717 — original branch included unrelated PTv3 commits. Reopening from jayakumarpujar:fix/landcover-val-oom-clean.
